src/transformers/modeling_utils.py

```diff
@@ -1718,6 +1718,23 @@ def save_pretrained(
             for ignore_key in self._keys_to_ignore_on_save:
                 if ignore_key in state_dict.keys():
                     del state_dict[ignore_key]
+        # Disable to see the damage.
+        if safe_serialization:
+            if self._keys_to_ignore_on_load_missing is not None:
+                from collections import defaultdict
+
+                ptrs = defaultdict(list)
+                for name, tensor in state_dict.items():
+                    ptrs[tensor.data_ptr()].append(name)
+
+                shared_ptrs = {ptr: names for ptr, names in ptrs.items() if len(names) > 1}
+
+                for _, names in shared_ptrs.items():
+                    for name in names:
+                        for pat in self._keys_to_ignore_on_load_missing:
+                            if re.search(pat, name):
+                                if name in state_dict:
+                                    del state_dict[name]
```

Comment:
This is roughly copy-pasted from `from_pretrained` just we drop only the shared keys (which allows dynamically unlinked tensors to go through as-is)